### PR TITLE
[clang] Bail out if the result of function template instantiation is not a function.

### PR DESCRIPTION
**Pull request by VitaNuo**

### CODE REVIEW

File: clang/lib/Sema/SemaTemplateInstantiate.cpp [hunk 1]
```diff
@@ -2649,7 +2649,9 @@ TypeSourceInfo *Sema::SubstFunctionDeclType(TypeSourceInfo *T,
   } else {
     Result = Instantiator.TransformType(TLB, TL);
   }
-  if (Result.isNull())
+  // When clang goes into recovery mode, it might substitute
```
Comment (sam-mccall) [inline clang/lib/Sema/SemaTemplateInstantiate.cpp:2652]: Not sure "recovery mode" is well-defined enough to refer to in this way

maybe:
When there are errors resolving types, clang may use IntTy as a fallback, breaking our assumption that function declarations have function types.

(I thought there was a name for this usage of IntTy but I can't find one. It happens e.g. at GetFullTypeForDeclarator, SemaType.cpp:5734)

File: clang/test/SemaTemplate/function-decl-nested-type-alias.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// RUN: %clang_cc1 -x c++ -std=c++14 -fsyntax-only -verify %s
+
+template <class A>
+using Type = typename A::NestedType; // expected-error {{type 'float' cannot be used prior to '::' because it has no members}}
+
+template <typename T>
+void Func() {
+  using MyType = Type<T>(); // expected-note{{in instantiation of template type alias 'Type' requested here}}
+  MyType var;
```
Comment (sam-mccall) [inline clang/test/SemaTemplate/function-decl-nested-type-alias.cpp:9]: // This is a function declaration, not a variable declaration!
// After substitution, we do not have a valid function type, and used to crash.

**Review by sam-mccall — APPROVED** [review]
Thanks! Looks great, just comment nits

**Comment by VitaNuo** [comment]
Pre-empted by https://github.com/llvm/llvm-project/pull/69459.

**Pull request closed without merging**